[JENKINS 60920] header breadcrumbs CSS update

fqueiruga opened the pull request:
See [JENKINS-60920](https://issues.jenkins-ci.org/browse/JENKINS-60920).

This PR includes the header and breadcrumb visual updates that came out of the work of the UX SIG.
Explanation for some technical decisions on the issue

_NOTE: To enable/disable the change easily without restarting, you can use the `/script` URL (typically http://localhost:8080/script in local dev mode) and type `hudson.Functions.UI_REFRESH=true` to enable the new UI or `hudson.Functions.UI_REFRESH=false` to disable it._


### Proposed changelog entries

* Major change: visual revamp of the layout and icons of the header bar and breadcrumbs. Instances with plugins that depend on details of the Jenkins layout (e.g. Simple Theme Plugin) may experience UI/layout problems.
* Major change: A new header color scheme can be enabled by setting the system property `jenkins.ui.refresh` to `true`.

### Proposed upgrade guidelines

Instances with plugins that depend on details of the Jenkins layout (e.g. Simple Theme Plugin with custom CSS themes) may experience UI/layout problems. Minor patches to the themes may be required

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@uhafner 
@timja 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

## Screenshots

Classic UI on desktop
<img width="999" alt="classic-ui-desktop" src="https://user-images.githubusercontent.com/5738588/73462804-a1eeaa00-437c-11ea-8d50-c43b13c44ae3.png">

New UI on desktop
<img width="999" alt="new-ui-desktop" src="https://user-images.githubusercontent.com/5738588/73462816-a87d2180-437c-11ea-9031-455d4c86f101.png">

Classic UI on mobile
<img width="429" alt="classic-ui-mobile" src="https://user-images.githubusercontent.com/5738588/73462858-b632a700-437c-11ea-8653-2cb527d89712.png">

New UI on mobile
<img width="429" alt="new-ui-mobile" src="https://user-images.githubusercontent.com/5738588/73462885-bcc11e80-437c-11ea-9bd9-cf483ed03aac.png">


## TODOs:

- [x] create a jelly helper for using the `<svg>` tag for sprite icons.
- [x] remove comments and port the _layout.jelly_ changes to the _html.jelly_ file.
- [x] migrate dimensions to rems. At least for the font-sizes.
- [x] fix for ie11
- [ ] Optional: migrate the breadcrumbs separator and dropdown menu selector icons to svgs.

fqueiruga committed:
New styles for breadcrumbs and warnings monitor

fqueiruga committed:
Adds the svgxuse polyfill for svg sprites on IE 11

fqueiruga committed:
Fixes and refactor for the new UI CSS code

fqueiruga committed:
Created new styles for the page header

- Used the material icons svg sprites to deliver the icons
- Icons are rendered using a <svg> tag

fqueiruga committed:
Create new base styles file and undo changes to the layout-common and style css files

fqueiruga committed:
Make the new header color and logo section opt-in

- The new UI can be toggled with the -Dui.enableNewUi=true flag

fqueiruga committed:
Rename and refactor the page-header css

- Rename the css class namespace from .main-header to .page-header
- Page header CSS moved into a modules folder and now included on the new-base-styles.css. This is done to avoid loading a separate file everytime.
- Extracted color variables to their own file

fqueiruga committed:
Updated localization for header and admin monitor

- The pageHeader template receives localization strings as props
- Missing translation for the admin monitor on the it, bg and pl locales

timja commented on core/src/main/resources/lib/layout/layout.jelly:
```diff
@@ -92,7 +92,7 @@ THE SOFTWARE.
 </j:if>
 <x:doctype name="html" />
-<html>
+<html class="${h.getSystemProperty('ui.enableNewUi') ? 'new-ui' : ''}">
```
we can probably come up with a better name for this, it would have been nice to have included this in the PR description as well 😉 

```suggestion
<html class="${h.getSystemProperty('jenkins.UI_REFRESH') ? 'new-ui' : ''}">
```

```suggestion
<html class="${h.getSystemProperty('jenkins.ui.refresh') ? 'new-ui' : ''}">
```

```suggestion
<html class="${h.getSystemProperty('jenkins.ui.refresh.2020') ? 'new-ui' : ''}">
```

timja replied:
we can probably come up with a better name for this, btw I missed this in the changelog section and had to go read the code, maybe call something like this out in the reviewers section 😄 

```suggestion
<html class="${h.getSystemProperty('jenkins.UI_REFRESH') ? 'new-ui' : ''}">
```

```suggestion
<html class="${h.getSystemProperty('jenkins.ui.refresh') ? 'new-ui' : ''}">
```

```suggestion
<html class="${h.getSystemProperty('jenkins.ui.refresh.2020') ? 'new-ui' : ''}">
```

timja commented:
Code looks fine to me, except the commented out code which is on the TODO list, will re-review when TODOs are completed

Tested locally with old and new UI on a mac 13".

batmat commented:
@fqueiruga test failures looks legit
![image](https://user-images.githubusercontent.com/223853/73493014-29a2db80-43b2-11ea-8588-e831e92e3eff.png)

fqueiruga committed:
Most measurements migrated to REMs

fqueiruga committed:
Update core/src/main/resources/lib/layout/layout.jelly

Co-Authored-By: Tim Jacomb <t.jacomb@kainos.com>

fqueiruga committed:
Code cleanup and apply new-ui toggle to the html.jelly file

fqueiruga committed:
Breadcrumb fixes

- Reverts breadcrumb link styling to use nested selectors instead of the .breadcrumbs__link class. The reason is that several plugins would break.
- Shifted margins so that breadcrumbs can be fullwidth in the absence of the auto refresh link

fqueiruga commented:
I have set up the feature flag but nothing else on the `html.jelly` file (https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/layout/html.jelly).
The reason is that I don't see any references to the header on that file. I'll do more changes if needed.

halkeye commented on war/src/main/webapp/images/jenkins-header-logo-new.svg:
```diff
@@ -0,0 +1,13 @@
+    <g id="CSS-phase-components" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+        <g id="Header-bar-component—Jenkins—Laptop" transform="translate(-19.000000, -14.000000)">
+            <g id="CSS-/-header-bar">
+                <image id="jenkins-logo" x="19" y="14" width="20.9444447" height="29" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAALYAAAD8CAYAAADaM14OAAAABGdBTUEAALGN5fIAKQAAQABJREFUeAHsXQWAFPX3f3t73QXcEZd0SrciooCJgdg/OzD/KnagmD9brB8YqIBi0opKl3Tn9RF3HNedu//Pm93Z/c7sbNzdXiA+PWbm2zP75s37vtTRv9CkT6BvYmLbSqOxj9Fg6G3U6TqR0RiOCcOIdOFGHHVGCjMSynTkgfJK+U+n86jW4dpoNFagLIfImIL+qXqDLkXnYUyp0+v ...
```
This is an SVG of a base64 encoded PNG

Seems like the least efficient way to store a png

fqueiruga replied:
Good catch. I changed it to a SVG. I also optimized it through a SVGO frontend

halkeye commented on core/src/main/resources/lib/layout/pageHeader.jelly:
```diff
@@ -0,0 +1,137 @@
+<!--
+The MIT License
+
+Copyright (c) 2004-2010, Sun Microsystems, Inc., Kohsuke Kawaguchi,
```
I don't think the copywriter should be added to this new file.

fqueiruga replied:
Removed!

halkeye commented on war/src/main/less/base/layout-commons.less:
```diff
@@ -0,0 +1,140 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
```
Same with this one. Where is it sources from that needs it to be specifically licensed from cloudbees

fqueiruga replied:
Removed

timja commented on war/src/main/webapp/images/jenkins-header-logo-new.svg:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
naming a file with `new` in it is generally a bad idea, it will likely stay forever.

possibly add `v2` ?

fqueiruga replied:
Renamed to `jenkins-header-logo-v2.svg`. I have also renamed other references to the "new ui". For example, the new css files are `base-styles-v2.less` and `ui-refresh-overrides.less`. LESS variables were also changed.

Please reopen this convo if you feel the changes weren't right

uhafner commented:
Thanks for providing the new styles as PR! It is much simpler to get an impression of the changes now... 

Before I am looking into the PR just a few quick impressions: on a plain vanilla Jenkins everything looks really nice! One small thing: If I disable the new UI (using the property) I still get the new breadcrumbs, is that intended? Or should everything be restored to the original state?

Did you try to use it with a theme? I'm using http://afonsof.com/jenkins-material-theme/ on my machine but these kind of themes seem not to work anymore. Even if I disable the new UI with the property, the theme looks weird.

imod commented:
I use the same theme as @uhafner - but to be honest, I don't think we should take to much care - it does not seem to be maintained anymore

uhafner commented:
Hmm, the simple them plugin has > 25000 installations. I'm not sure if it is only the material theme that will break, did you check one of the maintained themes like https://github.com/TobiX/jenkins-neo2-theme still works after the changes?

timja commented:
I thought it was expected that those plugins will have issues and will need to update? Possibly the need for those will go-away / decrease as the UI is improved / refreshed

fqueiruga committed:
Implemented some fixes from the feedback received

- Removed box shadow webkit prefix
- Fixed formatting on the breadcrumbs.css and layout-commons.less files
- Renamed the new UI toggle class from _new-ui_ to _ui-refresh_
- Renamed references to the _new ui_ to _v2_

fqueiruga commented on core/src/main/resources/lib/layout/pageHeader.jelly:
```diff
@@ -0,0 +1,137 @@
+                <img src="${imagesURL}/jenkins-header-logo-new.svg"
+                     alt="[${logoAlt}]"
+                     class="page-header__brand-image" />
+                <span class="page-header__brand-name">Jenkins</span>
```
Is there a i18n property for the product name?

alecharp replied:
No, Jenkins is Jenkins in all languages.

fqueiruga commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/resources.css:
```diff
@@ -32,7 +31,11 @@
-    box-shadow: 0 7px 20px rgba(0,0,0,.1);
+
+    /* Darken the box shadow to make the popup visible over the search box  */
+    -webkit-box-shadow: 0 1px 7px 0 rgba(0,0,0,0.3);
```
You're right. removed!

fqueiruga commented on core/src/main/resources/lib/layout/breadcrumbs.css:
```diff
@@ -18,53 +11,106 @@
+.breadcrumbs__auto-refresh:hover, .breadcrumbs__auto-refresh:focus {
+  /* TODO: change hover state
+     add it in the LI element so that it applies to the menuSeparator */
+    text-decoration: underline;
```
I have fixed the formatting for the whole file, all of it is 4 spaces now.

fqueiruga commented on war/src/main/less/base/layout-commons.less:
```diff
@@ -0,0 +1,140 @@
+ * THE SOFTWARE.
+ */
+
+ html {
```
you're right. fixed.

fqueiruga commented on core/src/main/resources/lib/layout/layout.jelly:
```diff
@@ -92,7 +92,7 @@ THE SOFTWARE.
 </j:if>
 <x:doctype name="html" />
-<html>
+<html class="${h.getSystemProperty('jenkins.ui.refresh') ? 'new-ui' : ''}">
```
Class renamed to `ui-refresh`.

fqueiruga commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/footer_de.properties:
```diff
@@ -22,3 +22,4 @@
 
 Manage\ Jenkins=Jenkins verwalten
 tooltip={0,choice,0#Keine Administrator-Warnungen sind|1#{0} Administrator-Warnung ist|1<{0} Administrator-Warnungen sind} aktiv.
+monitors={0,choice,0#Warnungen|1#Warnung|1<Warnungen}
```
@uhafner do you think this one's right? I tried to set the German translation based on the previous line (didn't try with the other translations for this file)

uhafner replied:
Yes that is correct!

fqueiruga commented:
Addressing your comments @uhafner 

The main agreement we got out of the SIG, based on the discussions motivated by @oleg-nenashev , was that changes to the layout and responsiveness would go straight into core. Only changes to the logo section and colors would be hidden behind the feature toggle. I can do further work on this if everybody feels it's needed.

On the topic of themes, both the material and neo2 theme work well enough with the UI fallback disabled: 

Material theme
<img width="1280" alt="jenkins-new-material-theme" src="https://user-images.githubusercontent.com/5738588/73759715-a21bea80-476c-11ea-9424-ce397805b239.png">
Neo2 theme
<img width="1277" alt="jenkins-new-neo-theme" src="https://user-images.githubusercontent.com/5738588/73759749-ac3de900-476c-11ea-9ac3-0959d04ab2d0.png">

The goal was to not break the themes intentionally, but also not holding back too much. Currently, only the search bar icons are presenting issues. @TobiX and @afonsof feel free to chime in.

TobiX commented:
> The goal was to not break the themes intentionally, but also not holding back too much. Currently, only the search bar icons are presenting issues.

Feel free to break Neo2 - I'll try to adapt the theme to the new design :D

fqueiruga committed:
Created jelly helper to use svg icons

fqueiruga commented:
Just created a jelly helper to use svg icons on https://github.com/jenkinsci/jenkins/pull/4463/commits/7751b3489bb533db92cba577cc9175bece45b41b

I tried to make it fully reusable and adaptable. Feedback appreciated.

fqueiruga committed:
Fix: restore proper icon for the user on the header

fqueiruga committed:
Fixes for Edge, IE 11 and IE 10

Wadeck commented:
<details>

<summary>1) "Menu arrow" hover</summary>

### Breadcrumb arrow hover
1a) The name and the arrow of a menu item are bound but "not naturally". For example the bottom and top of the arrow are part of the name and could lead to unexpected target when clicked.

![header_arrow_hover](https://user-images.githubusercontent.com/2662497/73864893-ac101d00-4842-11ea-9ed3-1d0c5c0773f5.gif)

Proposal: having a clear separation on the background hover effect to explicit the target of the current location.
![breadcrumb-name-hover](https://user-images.githubusercontent.com/2662497/73865482-9bac7200-4843-11ea-98bf-cfdb87fa35ab.png)
![breacrumb-arrow-hover](https://user-images.githubusercontent.com/2662497/73865488-9cdd9f00-4843-11ea-94f2-23a5814c9e36.png)

----

### Menu arrow hover
1b) The name and the arrow of the user are not bound like the breadcrumb. The area around the arrow makes the user name to be hover-ed but once on the arrow, the background effect disappears.

![user_arrow_hover](https://user-images.githubusercontent.com/2662497/73865597-cc8ca700-4843-11ea-930e-f7e7384d8262.gif)

Proposal: Separate the two like for the breadcrumb proposal

</details>

<details>

<summary>2) Header title</summary>

To have such a long area for the title seems misleading. It's easy to click by mistake on that large without the desire to come back to home page.
![header_title_long](https://user-images.githubusercontent.com/2662497/73865919-49b81c00-4844-11ea-874a-8cbb6dfe0d21.png)

Proposal: Having only the logo + title to be a link, the rest of the place to be inert.

</details>

<details>

<summary>3) Admin monitors</summary>

3a) Comparing the proposed and the current design, it seems the admin monitors are more easily visually ignored.

![current_am](https://user-images.githubusercontent.com/2662497/73866342-f85c5c80-4844-11ea-961f-20bf913b3dff.png)

![proposed_am](https://user-images.githubusercontent.com/2662497/73866345-f8f4f300-4844-11ea-9f33-41259a96df43.png)

Proposal: Either keep the same design idea as the existing design or 

----

3b) Once the warning popup is open, the white arrow from the popup to the menu item seems weirdly place.

![current_am_arrow](https://user-images.githubusercontent.com/2662497/73866648-7caedf80-4845-11ea-971b-4bc79b3eb69e.png)

![proposed_am_arrow](https://user-images.githubusercontent.com/2662497/73866651-7d477600-4845-11ea-984c-915009d3921b.png)

Proposal: Move it to the middle of the menu item.

----

3c) The notification icon used for the admin monitors does not bring the expected information. We expect this icon to inform the user he received a message or an event. Here the information is more about a "fix warning stable in the time".

Proposal: Use the warning or danger icon instead.

</details>

<details>

<summary>4) ui-refresh</summary>

4a) It feels weird to have the logo/title being "replicated" just below. In the other design, it does not feel like it because the "Jenkins" title seemed to be part of the logo instead of being a separate title.
![Screenshot_2020-02-05_183551_001](https://user-images.githubusercontent.com/2662497/73867155-79682380-4846-11ea-82be-947942b7747d.png)

Proposal: 
- Do not care
- Change font of the title to prevent the replication effect
- (difficult) Remove the first item in the breadcrumb on ui-refresh
- (bonus point) Merge the breadcrumb with the header, like in IntelliJ (proposal just below)

![Screenshot_2020-02-05_184119_001](https://user-images.githubusercontent.com/2662497/73867489-1cb93880-4847-11ea-9399-7be0dd0202e7.png)

----

4b) The diagonal separation feels weird / unnatural. I do not really understand its purpose. Yes it's a style, but it's not commonly spread to have diagonal separator like this. Especially as this one seems to be included in the breadcrumb, as they often have the / as separator.
_(In "Don't make me think", it's proposed to keep as much as possible the design / components that are common to prevent the user need to learn new things everytime)_ 

Proposal: Using this diagonal as the real breadcrumb separator
![proposal_breadcrumb](https://user-images.githubusercontent.com/2662497/73868571-fd230f80-4848-11ea-84de-03476df99233.png)

----

4c) The name of the items are capitalized. In terms of style it's nice, but in terms of usability I have some concerns. For the direct URL it's case-insensitive so it's fine, but for other usage of the job name, it could be problematic (not example at hand).

![current_folder_name](https://user-images.githubusercontent.com/2662497/73868746-525f2100-4849-11ea-9173-c4cf3efeadd2.png)

![proposed_folder_name](https://user-images.githubusercontent.com/2662497/73868748-52f7b780-4849-11ea-95fc-ceaae2fc4e94.png)

Proposal: keep the name as is to avoid any potential usability problem.

</details>

<details>

<summary>5) Search</summary>

The popup expand/diminish when the terms are typed.
![search_popup_move](https://user-images.githubusercontent.com/2662497/73869369-70795100-484a-11ea-807c-2326d709ffe6.gif)

Proposal: Do not correct in this PR, this "feature" was already present.

![search_popup_move_current](https://user-images.githubusercontent.com/2662497/73869364-6eaf8d80-484a-11ea-86f9-7ba1e81ef087.gif)

</details>

<details>

<summary>6) Misc</summary>

6a) In general it feels weird that for a single link, you have the hover effect on the background **and** the underline effect.
Proposal: One is sufficient, with a preference for the background hover effect.

----

6b) The height / font-size of the new header / breadcrumb seems huge compared to the past. In terms of screen size occupancy, I find it too big. We are coming from a 40px + 28px to a 56px + 40px header part.

Proposal: Reduce a bit the font-size, reduce the height of the breadcrumb.

----

6c) A text inside a div or a p is not rendered the same.

![am_warning_typography](https://user-images.githubusercontent.com/2662497/73866734-a6680680-4845-11ea-8deb-e56c5421b2c0.png)

Proposal: The typography should perhaps take care of a `<p>` or a `<span>` / `<div>` the same way in terms of style to avoid adding inconsistency in existing code.

</details>

Overall the new design is refreshing 🎉

fqueiruga commented:
Thank you for providing such comprehensive feedback Wadeck! Glad to hear you find it refreshing :D I'll try to answer it properly:

<details>
<summary>1 - Menu arrow hover</summary>
Most of the problematic behavior is already existing and has to do with the way the arrow is implemented in JS. We did not change this because we avoid changing JS and because that arrow is used in other UI elements, such side panels or tables :(.

_1a) Breadcrumb arrow hover:_ I'll try to fix the clicking underneath the arrow issue. Not much I can do about the background.
</details>

2 - Logo link: that's a bug, thanks for catching that one!

3 - Admin monitors: these feedback pieces are basically design changes and as such I think should be discussed on the SIG chat. We also. have iterated a lot on the design of this CTA and it seemed most people were fine with it.

<details>
<summary>4 - breadcrumbs</summary>

_4a) brand / breadcrumb confusion:_ you're not the first one to bring that up. The proposed solution was to rename the breadcrumb (here's the ticket for it https://issues.jenkins-ci.org/browse/JENKINS-60972). Some of those design changes could work. Please rise them on the SIG chat!
TBH I don't think the breadcrumbs belong under the header nor within it. I'd put them within the main panel.

_4b) diagonal border:_ this is mostly an style issue. What I understood from that book you mention is to not use stuff like `<select>` tags for navigation or stuff like that. 

_4c) breadcrumb capitalization:_ this is something we'll probably roll back, also because some languages like German give capitalization meaning.
</details>

5 - search: existing behavior and located in JS. It's tricky to change that's why we didn't even try.

<details>
<summary>6 - Misc</summary>

_6a) hover effects:_ I disagree, redundant visual cues improve the accessibility of the UI elements. We also cannot rely on color alone to convey information. The difference in luminosity may not be noticed by people with certain visual impairments and even on budget monitors.

_6b) header & breadcrumb height:__ 40px is commonly considered the minimum accesible size for interactive UI elements. The dimensions, 56px and 40px are not random and are sort of an industry standard. I don't find it to be a problem specially considering that the header hides on scroll.

_6c) text-sizing:_ we didn't work on the monitors popup itself because we are gonna tackle it on a future iteration! Also, I find that the `<p>` tag usually has it's own font size and margins but that's something we can talk about when revisiting typography.
</details>

**TLDR:** I'll start working on _2_ and _4c_. Not much I can do about _1_ but I'll try. There's a ticket open for _4a_. I think the design suggestions should be mentioned on the SIG chat, and I will hold off acting on them unless more people request it. We also can revisit the header in the future.

fqueiruga committed:
Improvements on the header and breadcrumbs

- Logo v2 changed and optimized using the svgo frontend https://jakearchibald.github.io/svgomg/
- Refactored the ammonitor insertion code
- Breadcrumbs are no longer capitalized
- Classic logo link no longer stretches past the images
- Fixed classic logo link on IE11
- Updated color of the footer
- Menu items background is also changed when hovering over the menu selector
- Removed copyright notice on top of some files

fqueiruga commented:
I think the main bulk of work has been finished. Could anyone please remove the WIP label?

fqueiruga commented:
I have diagnosed the reason for the test failures. 

The test cases ([this one for example](https://github.com/jenkinsci/jenkins/blob/820b64a6a6fce4ed6661a8e38ad5905dee3db16f/test/src/test/java/hudson/model/UserTest.java#L147)) try to verify that all images on the page load.

To do it, it uses [this function](https://github.com/jenkinsci/jenkins-test-harness/blob/93e9b64a89d3048896f77d62fba84fff715b932b/src/main/java/org/jvnet/hudson/test/JenkinsRule.java#L1513) from jenkins-test-harness. My understanding is that the function verifies that all images referenced on `<img>` tags have valid heights. This works fine on PNGs but fails on SVGs, as they don't seem to have effective heights.

Any advice on how to act on this? What comes to mind:
- Skip that test case 
- Update the jenkins-test-harness library
- Copy-paste the helper code from jenkins-test-harness to this project and then modify it to ignore SVGs?

timja commented:
Does it make sense to remove the WIP label while the tests are failing?

why would it fail on svg's? Aren't they using their own tag, and it's selecting img tags?

the jenkins-test-harness should be updated if an issue is found there, an incremental release should then be automatically produced on the pull request and then can be consumed here to verify it fixes the issue.

then we can merge the fix to jenkins-test-harness, release it and consume it here.

fqueiruga commented:
I suggested to remove it because right now I'll only be addressing feedback and fixing tests. It's OK for me to keep it. 

The tests are failing on `<img>` tags that have an svg source (the Jenkins logos), not on icons.

I'll look into addressing this on the jenkins-test-harness.

MarkEWaite commented:
I removed the work-in-progress label since merging will be blocked until the checks pass.  We don't need to hold the pull request in "double jeopardy".  One barrier to merge is enough.

Wadeck commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/resources.css:
```diff
@@ -1,29 +1,28 @@
+
+#visible-am-button .am-monitor__count {
+    display: inline-block;
+    display: inline-flex;
```
Could you put a comment explaining why you put the two there?

I imagine it's due to the potential lack of support for the inline-flex but in  https://caniuse.com/#feat=flexbox, it seems that all the supported browsers are ok with this.

fqueiruga replied:
I just do it to be extra defensive. In my experience I learned to be extra paranoid when adding flexbox fallbacks. Once we add postcss most of this can be avoided. 

I can undo these changes if you prefer.

Wadeck replied:
> I can undo these changes if you prefer.

Now that we have the explanation, if someone dig into the code they will find this comment and understand. That's good.

Wadeck commented:
> 1 - Menu arrow

Good point, just tested with ci.jenkins.io and the behavior is the same, even with the "bottom the arrow" stuff. Do not worry correcting this stuff right now ;)

---

> Admin monitors: these feedback pieces are basically design changes and as such I think should be discussed on the SIG chat. We also. have iterated a lot on the design of this CTA and it seemed most people were fine with it.

Concerning the point 3b, it seems more like a ui-bug than something expected.

For the 3a, TBH I am a bit confused about the answer.

⚠️ @oleg-nenashev @daniel-beck do you feel confident to publish this visibility change (reduction from my PoV) for the security warnings?

FTR for non-ui aware, CTA = call-to-action.
To keep the inertia and not having this PR blocked due to a single element, it could be interesting to have the admin monitor change in a separate PR.

---

> 6a) hover effects: I disagree, redundant visual cues improve the accessibility of the UI elements.

I would say it's non-standard. As Jenkins is developer-centric, one can look at how GitHub or Jira behave around that point, lots of different (and inconsistent) hover effect, but I was not able to find any redundant among them. Nothing blocking here, just a bit weird ;)

Wadeck reviewed (commented):
> 1 - Menu arrow

Good point, just tested with ci.jenkins.io and the behavior is the same, even with the "bottom the arrow" stuff. Do not worry correcting this stuff right now ;)

---

> Admin monitors: these feedback pieces are basically design changes and as such I think should be discussed on the SIG chat. We also. have iterated a lot on the design of this CTA and it seemed most people were fine with it.

Concerning the point 3b, it seems more like a ui-bug than something expected.

For the 3a, TBH I am a bit confused about the answer.

⚠️ @oleg-nenashev @daniel-beck do you feel confident to publish this visibility change (reduction from my PoV) for the security warnings?

FTR for non-ui aware, CTA = call-to-action.
To keep the inertia and not having this PR blocked due to a single element, it could be interesting to have the admin monitor change in a separate PR.

---

> 6a) hover effects: I disagree, redundant visual cues improve the accessibility of the UI elements.

I would say it's non-standard. As Jenkins is developer-centric, one can look at how GitHub or Jira behave around that point, lots of different (and inconsistent) hover effect, but I was not able to find any redundant among them. Nothing blocking here, just a bit weird ;)

Wadeck commented on war/src/main/less/modules/page-header.less:
```diff
@@ -0,0 +1,226 @@
+
+.page-header__brand-image {
+    height: 32px;
+    height: 2rem;
```
Same comment as above with https://caniuse.com/#search=rem

fqueiruga replied:
The reason I did this for the logo is to have an extra piece of mind. Without this, in really old browsers the logo would go fullscreen. That way, you wouldn't even see a faulty page.

fqueiruga commented:
Answering your points @Wadeck 

3) Activity monitor:

- Regarding 3a, what I meant to answer is that we did discuss and performed design iterations on this very component, and nobody raised an issue with it's visibility. I agree to create a separate discussion to determine if this UI piece needs further work, and see if more people share these concerns.
- On 3b, the little arrow indicator. We can definitely remove it if it feels redundant.

6a) Hover effects:
Jenkins being developer-centric does not invalidate the need for visual accessibility improvements. There are definitely devs with poor eyesight / budget monitors that can benefit from these practices. Also, those products do implement them (look at the PR links on the Github dashboard, they change both underline and link color).

fqueiruga commented:
I created a PR to fix JTH on https://github.com/jenkinsci/jenkins/pull/4463

fqueiruga committed:
Bumped JTH version with the SVG download fix

fqueiruga commented:
@timja It seems that the tests pass using the incremental version of JTH generated on https://github.com/jenkinsci/jenkins-test-harness/pull/200. What would be the next step here?

timja commented:
Someone with merge/ release permission to take a look at https://github.com/jenkinsci/jenkins-test-harness/pull/200

oleg-nenashev reviewed (changes requested):
Thanks for the pull request @fqueiruga ! I am fine with the change as long as the new color scheme and the logo layout remain opt-in. I might be a bit conservative, but I still like the current header more TBH (especially the left part)... But it is fine to experiment.

I am requesting change, because `h.getSystemProperty()` uses the String API of `SystemProperties`. At the same time, there are Boolean APIs which are designed for the use-case you are trying to achieve: SystemProperties . I highly recommend switching to Boolean API and introducing new bridge methods in `Functions` if needed.

Also keep in mind that `SystemProperties` API is rather designed for system properties which values are calculated once on restart. It involves calls like `ServletContext` checks, and I would not guarantee that it returns results immediately in all cases. Be careful when using the live switch in the frontend. If it is really needed, maybe using a `private static boolean` cache is preferable

oleg-nenashev commented:
JTH patch was released: https://github.com/jenkinsci/jenkins-test-harness/releases/tag/jenkins-test-harness-2.61

timja commented on test/pom.xml:
```diff
@@ -70,7 +70,7 @@ THE SOFTWARE.
       <groupId>${project.groupId}</groupId>
       <artifactId>jenkins-test-harness</artifactId>
-      <version>2.59</version>
+      <version>2.61-rc1313.7db210c24195</version>
```
```suggestion
      <version>2.61</version>
```

fqueiruga committed:
Update test/pom.xml

Co-Authored-By: Tim Jacomb <t.jacomb@kainos.com>

fqueiruga committed:
Added a function to check boolean system properties from jelly

fqueiruga commented:
@oleg-nenashev I have created a function to check the boolean system properties and added a test for it.

I understand that this approach can present issues if used with a live switch. If we implement a user toggle for the new UI we should probably move to a better approach.

fqueiruga commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -493,6 +493,16 @@ public static String getSystemProperty(String key) {
+     * Delegates to {@link SystemProperties#getBoolean(java.lang.String)}
+     */
+    @Restricted(DoNotUse.class)
+    public static boolean isSystemPropertyEnabled(String key) {
```
Not sure about the name. Opinions?

timja replied:
sounds fine to me

daniel-beck replied:
Doesn't really matter for private API.

fqueiruga committed:
Fix header links and add svg styles for legacy layout

fqueiruga commented on war/src/main/webapp/css/style.css:
```diff
@@ -137,7 +137,7 @@ dt > .plugin_version {
 }
 
-#header .login {
+/* #header .login {
```
Actually this was a bug, I have uncommented it. This file is the legacy styles, and I kept it to support the warnings-ng plugin until it's migrated (it depends on this file and the layout-commons.css).

I have also backported the `.svg-icon` to this file, for the same reason.

fqueiruga commented:
Just to summarize the status. I re-requested reviews because all feedback pieces have been addressed.

oleg-nenashev reviewed (approved):
All my concerns are addressed. I like how the default Web UI looks after the changes, and I am fine with merging this pull request. Experimental UI is experimental for a reason, so we can keep tweaking it after the merge

daniel-beck commented:
Since I got pinged by others about the visual de-emphasizing of the admin monitors, is there a reference doc to help in understanding the nontechnical decisions that resulted in this design?

fqueiruga committed:
Some bug fixing for the new header

- The ui-refresh changes were not properly being enabled with the boolean system property. Added explicit true check on the jelly templates.
- Added an explicit width for the ui-refresh logo because it would mess with the layout

fqueiruga commented:
@daniel-beck there were conversations on the UX SIG slack channel. It was mostly a matter of adding an icon and a text copy, and keeping the look & feel consistent.

Wadeck reviewed (approved):
Not happy with the admin monitor part but I will not block the PR for that. 

<details>

<summary>Search focus</summary>

#### Effect on "classic" UI
![search-focus-border_current](https://user-images.githubusercontent.com/2662497/74724468-9801ed80-523c-11ea-89b4-0e5d7dc6a157.gif)
=> Nice effect

#### Effect on "experimental" UI
![search-focus-border_experimental](https://user-images.githubusercontent.com/2662497/74724508-a2bc8280-523c-11ea-81f6-90677baf696a.gif)
=> seems against the accessibility objective of the approach no? The contrast does not seem sufficient. But it's an experimental feature, so it could be polished afterwards.

</details>

<details>

<summary>Arrow menu</summary>

#### After change on the arrow menu in the header
![arrow-menu-header](https://user-images.githubusercontent.com/2662497/74725014-8b31c980-523d-11ea-897c-b76cdd41fad4.gif)
There is no longer a separation between the two features, in terms of UI consistency it's better but for the usability it's a bit worse than before I feel.

</details>

⚠️ As https://github.com/jenkinsci/jenkins/pull/4503 is ready-for-merge since ~23h, it seems interesting to adapt this PR once the other is merged. 

⚠️ And due to that, having two features called "refresh" could be confusing. As a user without internal knowledge about the PRs, when I will see the experimental feature flag for this UI revamp (`jenkins.ui.refresh`), I am not sure I will be able to know it's about the new UI or the refresh removal.

Not blocking the PR, up to you to correct the points or not

fqueiruga commented on war/src/main/webapp/css/style.css:
```diff
@@ -58,12 +58,12 @@ body, table, form, input, td, th, p, textarea, select
 @media (min-width:1600px){
   body#jenkins.j-hide-left #main-panel{max-width:75%}
   body, table, form, input, td, th, p, textarea, select, #tasks .task
-    {font-size:14px;}  
```
Resolving this one. You're right but I don't want to mess things up by adding more unneeded commits

daniel-beck commented:
Some feedback from testing this below. Very minor findings, nonblocking.

---

Clicking "monitor" twice (to close the popup again) will result in the "monitor" button still being focused, including underline. This is the same behavior as the current button has, but since the focus indicator is more pronounced than just a slightly different shade of red, it seems weird now.

> ![Screenshot](https://user-images.githubusercontent.com/1831569/74728699-86701400-5243-11ea-8773-36904f9e2097.png)

---

At fractional window sizes between 767 and 768 pixels wide (yay page zoom), the top bar shows labels and search box but isn't wide enough to support that. It's not something that matters, obviously, but potentially indicative of a real bug?

> ![Screenshot](https://user-images.githubusercontent.com/1831569/74728487-25484080-5243-11ea-93e9-6116ef9288f8.png)

fqueiruga commented:
To Wadeck:

- Maybe you can raise those design issues in a future SIG meeting, all components are open for several iterations.
- The click area for the dropdown menu is the same as before, 16 pixels wide, maybe it looks smaller because the font is bigger.
- You do have a point regarding the feature toggle name being confusing, but I think it's sort of a bigger change that I'm comfortable now.

To Daniel:
- The focus state being kept is a known Chrome and it happens with almost all clickable elements. Not much we can do on our side :(
- The breakpoints issue is a nice find. I used the css classes provided already existing in Jenkins and they bring this behaviour. I can file a bug report and fix this in a future PR if that's OK

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -493,6 +493,16 @@ public static String getSystemProperty(String key) {
+     */
+    @Restricted(DoNotUse.class)
+    public static boolean isSystemPropertyEnabled(String key) {
+        return SystemProperties.getBoolean(key);
```
Oleg in https://github.com/jenkinsci/jenkins/pull/4463#pullrequestreview-357490289:

> Also keep in mind that `SystemProperties` API is rather designed for system properties which values are calculated once on restart. It involves calls like `ServletContext` checks, and I would not guarantee that it returns results immediately in all cases. Be careful when using the live switch in the frontend. If it is really needed, maybe using a `private static boolean` cache is preferable

fqueiruga replied:
This was done to address the concern that it was a boolean property. We understand this approach is not optimal for doing live switches. Of course, we will choose a technique that addresses these concerns once we start working on a UI-based toggle.

fqueiruga replied:
Resolving this one, feel free to reopen it if you think it deserves further discussion

daniel-beck replied:
@fqueiruga My point is that the toggle should be implemented via

```
@Restricted(NoExternalUse.class)
public static /* non-final */ boolean UI_REFRESH = SystemProperties.getBoolean("jenkins.ui.refresh");
// and optionally a getter for it if accessing code can be made nicer
```

This way, you're not evaluating the potentially expensive operation on every page render but only once on startup, while retaining the ability to edit it at runtime (by setting the static var).

fqueiruga replied:
I understand now. If it's not a blocker I'd rather follow up on this on a future PR rather than change this again, as it's a risky one. Feel free to reopen this convo again if you think it's necessary.

daniel-beck replied:
@fqueiruga Addressed in https://github.com/jenkinsci/jenkins/pull/4463/commits/205fa3a7590439f63e8059fcee086614386f7b2e

(For which I got Felix's approval out of band)

daniel-beck commented:
The admin monitor looks a bit weird for two reasons:

* We retain pluralization, so I guess it is intended to say something like "1 monitor" or "3 monitors", but between the label and the stylized number, this is pretty unclear. The order of elements makes this especially weird, because I would expect "monitors: 1" in this case (i.e. always plural).
* The red text color is barely visible and even looks like there's something wrong with the number text. a more pronounced color (GitHub uses blue for notifications), or coloring the background rather than the number might work better here?

I do not feel strongly about a de-emphasizing of monitors. Security warnings are some of them, but for from the only ones. While not as readily noticeable like a giant red block, the width of the thing alone makes it still noticeable _enough_ I think (and could always be iterated on).

Have not signed up to the CDF Slack, so may lack context. Would consider my feedback here to be generally nonblocking.

fqueiruga commented:
Thanks Daniel, I'll surface these pieces of design feedback on the next SIG meeting. I find them interesting, especially the pluralization one.

alecharp commented on core/src/main/resources/lib/layout/svgIcon.jelly:
```diff
@@ -0,0 +1,49 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:d="jelly:define">
+    <st:documentation>
+        <!-- TODO: @since? -->
```
I don't think we put those on `jelly` files.

fqueiruga replied:
It's used on about 7-8 jelly templates. This is one example:   https://github.com/jenkinsci/jenkins/blob/dc8781dd69dc6fed1b95ea0837e5aa74982c8975/core/src/main/resources/lib/layout/css.jelly#L30 . I'd rather keep it.

timja commented:
@daniel-beck ux-sig is now on gitter https://gitter.im/jenkinsci/ux-sig

alecharp reviewed (approved):
Less files are pretty large, difficult to be sure everything is perfect, but it seems to be working fine. Just a few comments, no blockers.

fqueiruga committed:
Update core/src/main/java/hudson/Functions.java

Co-Authored-By: Adrien Lecharpentier <adrien.lecharpentier@gmail.com>

fqueiruga commented on core/src/main/resources/lib/layout/breadcrumbs.css:
```diff
@@ -1,70 +1,115 @@
+    font-weight: bold;
+
+    display: inline-block;
+    display: inline-flex;
```
It´s inline-flex but I left inline-block as a fallback.

Wadeck replied:
I think my initial comment proposing to add a comment in the code seems even more accurate :)

fqueiruga commented on core/src/main/resources/lib/layout/layout.jelly:
```diff
@@ -156,6 +156,9 @@ THE SOFTWARE.
     <link rel="stylesheet" href="${resURL}/scripts/yui/menu/assets/skins/sam/menu.css" type="text/css" />
     <!--link rel="stylesheet" href="${resURL}/scripts/yui/editor/assets/skins/sam/editor.css" type="text/css" /-->
 
+    <link rel="stylesheet" href="${resURL}/css/google-fonts/roboto/css/roboto.css" type="text/css" />
```
Maybe. Next iteration is typography so we can take the chance to refactor it then

oleg-nenashev reviewed (approved):
Taking the approvals, I suggest to merge it into 2.222 if there is no negative feedback by Thursday 4PM UTC

fqueiruga commented on core/src/main/resources/lib/layout/breadcrumbs.css:
```diff
@@ -80,21 +125,21 @@
-  visibility: hidden;
-  cursor: pointer;
-  z-index: 2000;
+    background-color:transparent;
```
Do you mean the indentation changes? The file indentation was a bit inconsistent as it had different indentation levels. I put all the lines on the same level.

fqueiruga commented on core/src/main/resources/lib/layout/breadcrumbs.css:
```diff
@@ -1,70 +1,115 @@
-    color:#3465a4;
+#breadcrumbs LI A:hover, #breadcrumbs LI A:focus,
+.breadcrumbs__auto-refresh:hover, .breadcrumbs__auto-refresh:focus {
+    /* TODO: change hover state
```
Yes, probably once we work on the contextual menu dropdown.

daniel-beck committed:
Do not check the system property on every page render

daniel-beck commented:
This PR introduces a minor regression on Firefox:

When jumping to an anchor on the page (e.g. through the dropdown menu for configuration forms), the breadcrumb bar doesn't move to the top of the page. Might be related to page zoom.

### Expected result:

> ![Screenshot](https://user-images.githubusercontent.com/1831569/75075820-f3630280-54fe-11ea-985f-612d691d6a1b.png)

### Actual result:

> ![Screenshot](https://user-images.githubusercontent.com/1831569/75075812-ee9e4e80-54fe-11ea-9035-cffee2f6bfc0.png)

daniel-beck commented:
This PR introduces a minor regression on Firefox:

When jumping to an anchor on the page (e.g. through the dropdown menu for configuration forms), the breadcrumb bar doesn't move to the top of the page. Might be related to page zoom.

### Expected result:

> ![Screenshot](https://user-images.githubusercontent.com/1831569/75075820-f3630280-54fe-11ea-985f-612d691d6a1b.png)

### Actual result:

> ![Screenshot](https://user-images.githubusercontent.com/1831569/75075812-ee9e4e80-54fe-11ea-9035-cffee2f6bfc0.png)

oleg-nenashev commented:
@fqueiruga FYI I added a short upgrade guide entry to address the comments in https://groups.google.com/d/msg/jenkinsci-dev/VRxRCP_IKhQ/8Uw2SJqWAAAJ . It is based on your changelog entry text. Please feel free to extend it as needed

CptPlastic commented:
> Hmm, the simple theme plugin has > 25000 installations. I'm not sure if it is only the material theme that will break, did you check one of the maintained themes like https://github.com/TobiX/jenkins-neo2-theme of @TobiX still works after the changes?

I am in the same boat as many that use the material theme. When you try to enable this it just breaks the login page. Even if you remove the plugin some thing still lingers. Is there any guidance for people who do run into this because we are back to the classic theme and people notice. Is there a place I can go look to fix this for the folks that have Simple Theme installed?

timja commented:
> > Hmm, the simple theme plugin has > 25000 installations. I'm not sure if it is only the material theme that will break, did you check one of the maintained themes like [TobiX/jenkins-neo2-theme](https://github.com/TobiX/jenkins-neo2-theme) of @TobiX still works after the changes?
> 
> I am in the same boat as many that use the material theme. When you try to enable this it just breaks the login page. Even if you remove the plugin some thing still lingers. Is there any guidance for people who do run into this because we are back to the classic theme and people notice. Is there a place I can go look to fix this for the folks that have Simple Theme installed?

each theme will need some updates done

i.e. material theme here:
https://github.com/afonsof/jenkins-material-theme

CptPlastic commented:
> > > Hmm, the simple theme plugin has > 25000 installations. I'm not sure if it is only the material theme that will break, did you check one of the maintained themes like [TobiX/jenkins-neo2-theme](https://github.com/TobiX/jenkins-neo2-theme) of @TobiX still works after the changes?
> > 
> > 
> > I am in the same boat as many that use the material theme. When you try to enable this it just breaks the login page. Even if you remove the plugin some thing still lingers. Is there any guidance for people who do run into this because we are back to the classic theme and people notice. Is there a place I can go look to fix this for the folks that have Simple Theme installed?
> 
> each theme will need some updates done
> 
> i.e. material theme here:
> https://github.com/afonsof/jenkins-material-theme

You see how that causes an issue though right? 25000 things have this and that is a pretty substantial impact. Even if they install that plugin and remove it the break still happens. The only way to unload it is to restart the service twice. I get that this plugin that is used and some of the templates need updates but this breaks jenkins bottom line. There may be something jenkins can do to force the css thats for some reason making the login pages blank. 

I think this is awesome and I would much rather have a nice UI that matched blue ocean and that you guys are giving some much needed attention to the UI I just wanted to make you aware that these things are happening.

uhafner commented:
> i.e. material theme here:
https://github.com/afonsof/jenkins-material-theme

The material theme seems to be unmaintained currently. So if nobody steps in this theme will not be supported in the near future.
